include/hwloc/rsmi.h

```diff
@@ -0,0 +1,214 @@
+#endif
+  return 0;
+}
+
+/** \brief Get the hwloc OS device object corresponding to the
+ * AMD GPU device whose index is \p dv_ind.
+ *
+ * Return the OS device object describing the AMD GPU device whose
+ * index is \p dv_ind. Returns NULL if there is none.
+ *
+ * The topology \p topology does not necessarily have to match the current
+ * machine. For instance the topology may be an XML import of a remote host.
+ * I/O devices detection and the RSMI component must be enabled in the topology.
+ *
+ * \note The corresponding PCI device object can be obtained by looking
+ * at the OS device parent object (unless PCI devices are filtered out).
+ */
+static __hwloc_inline hwloc_obj_t
+hwloc_rsmi_get_device_osdev_by_index(hwloc_topology_t topology, uint32_t dv_ind)
+{
+  hwloc_obj_t osdev = NULL;
+  while ((osdev = hwloc_get_next_osdev(topology, osdev)) != NULL) {
+    if (HWLOC_OBJ_OSDEV_GPU == osdev->attr->osdev.type
+      && osdev->name
+      && !strncmp("rsmi", osdev->name, 4)
+      && atoi(osdev->name + 4) == (int) dv_ind)
+      return osdev;
+  }
+  return NULL;
+}
+
+/** \brief Get the hwloc OS device object corresponding to AMD GPU device,
+ * whose index is \p dv_ind.
+ *
+ * Return the hwloc OS device object that describes the given
+ * AMD GPU, whose index is \p dv_ind Return NULL if there is none.
+ *
+ * Topology \p topology and device \p dv_ind must match the local machine.
+ * I/O devices detection and the RSMI component must be enabled in the topology.
+ * If not, the locality of the object may still be found using
+ * hwloc_rsmi_get_device_cpuset().
+ *
+ * \note The corresponding hwloc PCI device may be found by looking
+ * at the result parent pointer (unless PCI devices are filtered out).
+ */
+static __hwloc_inline hwloc_obj_t
+hwloc_rsmi_get_device_osdev(hwloc_topology_t topology, uint32_t dv_ind)
+{
+  hwloc_obj_t osdev;
+  rsmi_status_t ret;
+  rsmiPciInfo_t pci;
+  uint64_t id;
+  char uuid[64];
+
+  if (!hwloc_topology_is_thissystem(topology)) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  ret = rsmi_dev_pci_id_get(dv_ind, &(pci.bdfid));
+  if (RSMI_STATUS_SUCCESS != ret) {
+    errno = EINVAL;
+    return -1;
+  }
+
+  ret = rsmi_dev_unique_id_get(dv_ind, &id);
+  sprintf(uuid, "%xll", id);
+  if (RSMI_STATUS_SUCCESS != ret)
+    uuid[0] = '\0';
+
+  osdev = NULL;
+  while ((osdev = hwloc_get_next_osdev(topology, osdev)) != NULL) {
+    hwloc_obj_t pcidev = osdev->parent;
+    const char *info;
+
+    if (strncmp(osdev->name, "rsmi", 4))
+      continue;
+
+    if (pcidev
+      && pcidev->type == HWLOC_OBJ_PCI_DEVICE
+      && pcidev->attr->pcidev.domain == pci.domain
+      && pcidev->attr->pcidev.bus == pci.bus
+      && pcidev->attr->pcidev.dev == pci.device
+      && pcidev->attr->pcidev.func == pci.function)
+      return osdev;
+
+    info = hwloc_obj_get_info_by_name(osdev, "AMDUUID");
+    if (info && !strcmp(info, uuid))
+      return osdev;
+  }
+
+  return NULL;
+}
+
+/** @} */
```

Comment:
There was some idea of moving some of these helpers to somewhere else. If it's not a priority for you, I'd rather keep them in a separate commit so that we can merge the core backend and discuss helpers laters.

Reply:
agree